std/algorithm/searching.d

```diff
@@ -3462,6 +3462,181 @@ unittest
 }
 
 /**
+Computes the index of the first occurrence of `range`'s minimum element.
+
+Params:
+    pred = The ordering predicate to use to determine the minimum element.
+    range = The input range to search.
+
+Returns:
+    The index of the first encounter of the minimum element in `range`. If the
+    range is empty, -1 is returned.
+ */
+sizediff_t minIndex(alias pred = "a < b", Range)(Range range)
+    if (isForwardRange!Range && !isInfinite!Range &&
+        is(typeof(binaryFun!pred(range.front, range.front))))
+{
+    if (range.empty) return -1;
+
+    sizediff_t minPos = 0;
+
+    static if (isRandomAccessRange!Range && hasLength!Range)
+    {
+        foreach (i; 1 .. range.length)
+        {
+            if (binaryFun!pred(range[i], range[minPos]))
+            {
+                minPos = i;
+            }
+        }
+    }
+    else
+    {
+        sizediff_t curPos = 0;
+        Unqual!(typeof(range.front)) min = range.front;
+        for (range.popFront(); !range.empty; range.popFront())
+        {
+            ++curPos;
+            if (binaryFun!pred(range.front, min))
+            {
+                min = range.front;
+                minPos = curPos;
+            }
+        }
+    }
+    return minPos;
+}
+
+///
+@safe pure nothrow unittest
+{
+    int[] a = [2, 3, 4, 1, 2, 4, 1, 1, 2];
+    // Minimum is 1 and first occurs in position 3
+    assert(a.minIndex == 3);
+    // Get maximum index with minIndex
+    assert(a.minIndex!"a > b" == 2);
+
+    int[] b;
+    // Range is empty, so return value is -1
+    assert(b.minIndex == -1);
+
+    const int[] c = [2, 5, 4, 1, 2, 3];
+    // Works for const ranges too
+    assert(c.maxIndex == 1);
+    // Lambda
+    assert(c.minIndex!((a, b) => a < b) == 3);
```

Comment:
How about using a custom type here - could be a lot less confusing and closer to the real world (to flip the predicate one would use max):

```d
struct Dog { int age; }
auto dogs = [Dog(10), Dog(5), Dog(15)];
assert(dugs.minIndex!`d1.age < d2.age` == 3);
```

Note that not everyone is a friend of string lambda, but imho they are awesomely concise and thus well fitted for short example snippets.